cli/generate.go

```diff
@@ -0,0 +1,279 @@
+	clExtraFlags      *[]string
+	vlExtraFlags      *[]string
+	relayURL          string
+	mevBoostUrl       string
+	executionApiUrl   string
+	executionAuthUrl  string
+	consensusApiUrl   string
+}
+
+func GenerateCmd(sedgeAction actions.SedgeActions) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "generate [flags]",
+		Short: "Generate new setups according to selected options",
+		Long: `Generate new setups according to selected options and flags.
+
+It will create a 'docker-compose.yml' and an '.env', you will need later to run the docker-compose script.
+
+You can generate:
+- Full Nodes (consensus+execution+validator)
+- Full Nodes without Validator (consensus+execution)
+- Consensus Node
+- Execution Node
+- Validator Node
+- MevBoost Node
+`,
+		Args: cobra.NoArgs,
+	}
+
+	cmd.AddCommand(FullNodeSubCmd(sedgeAction))
+	cmd.AddCommand(ExecutionSubCmd(sedgeAction))
+	cmd.AddCommand(ConsensusSubCmd(sedgeAction))
+	cmd.AddCommand(ValidatorSubCmd(sedgeAction))
+	cmd.AddCommand(MevBoostSubCmd(sedgeAction))
+
+	cmd.PersistentFlags().StringVarP(&generationPath, "path", "p", configs.DefaultDockerComposeScriptsPath, "docker-compose scripts generation path")
+	cmd.PersistentFlags().StringVarP(&network, "network", "n", "mainnet", "Target network. e.g. mainnet, goerli, sepolia, etc.")
+	cmd.PersistentFlags().StringVar(&logging, "logging", "json", fmt.Sprintf("Docker logging driver used by all the services. Set 'none' to use the default docker logging driver. Possible values: %v", configs.ValidLoggingFlags()))
+	return cmd
+}
+
+func validateNetwork(network string) error {
+	networks, err := utils.SupportedNetworks()
+	if err != nil {
+		return fmt.Errorf(configs.NetworkValidationFailedError, err)
+	}
+	if !utils.Contains(networks, network) {
+		return fmt.Errorf(configs.UnknownNetworkError, network)
+	}
+	return nil
+}
+
+func preValidationGenerateCmd(network, logging string) error {
+	// Validate network
+	err := validateNetwork(network)
+	if err != nil {
+		return err
+	}
+
+	// Validate Logging flag
+	if err = configs.ValidateLoggingFlag(logging); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func runGenCmd(out io.Writer, flags *GenCmdFlags, sedgeAction actions.SedgeActions, services []string) error {
+
+	// Warn if exposed ports are used
+	if flags.mapAllPorts {
+		log.Warn(configs.MapAllPortsWarning)
+	}
+
+	// Warn if checkpoint url used
+	if flags.checkpointSyncUrl != "" {
+		log.Warnf(configs.CheckpointUrlUsedWarning, flags.checkpointSyncUrl)
+	}
+
+	// Get all clients: supported + configured
+	c := clients.ClientInfo{Network: network}
+	clientsMap, errs := c.Clients(lessMevBoost(services))
+	if len(errs) > 0 {
+		return errs[0]
+	}
+
+	// Handle selection and validation of clients
+	combinedClients, err := valClients(clientsMap, flags, services)
+	if err != nil {
+		return err
+	}
+
+	err = initGenPath(generationPath)
+	if err != nil {
+		return err
+	}
+
+	if flags.jwtPath, err = generateJWTSecret(flags.jwtPath); err != nil {
```

Comment:
comment this

Reply:
Done at faacf787d646dade513226be2384704e0e6f9ad9